### ISSUE TITLE
Ultimate Psionics - Persistent Focus (feat... is missing)

### ISSUE THREAD

**Issue by shadowchemosh** [shadowchemosh/HL-Pack-Pathfinder#6]
_From @Ualaa on March 29, 2015 16:46_

The feat 'Persistent Focus (Psionic)' is missing.  There doesn't seem to be any scripting, so I can likely manually add it.  At some point, it would be nice to have it included in the package.

Persistent Focus (Psionic)
You are treated as maintaining psionic focus for one ability.
Prerequisites: Wis 13, Psionic Meditation, Autohypnosis 4 ranks.
Benefit: Choose one psionic ability or psionic feat that requires maintaining psionic focus.  From this point forward, you are always considered psionicially focused even if you expend your psionic focus for any other ability.

_Copied from original issue: shadowchemosh/HL-Pack-3PP#41_

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 16:56_

been doublechecking the feats but only up to D now. :o)
can skip up to that one though

**Comment by shadowchemosh** [comment]
We can add the feat.  The script logic is going to take some work to implement.  All abilities, feats, and powers are looking at the "hero" for the tag Hero.PsiFocus.  Now we will need to have some other way for each "specific" Thing to check itself for a special tag.  Maybe "Psionics.PsiFocus"?

Then every script that is using "Hero.PsiFocus" will need it script logic changed from:
<strong>doneif (hero.tagis[Hero.PsiFocus] = 0)</strong>
to 
<strong>doneif (hero.tagis[Hero.PsiFocus] + tagis[Psionics.PsiFocus] = 0)</strong>

Then you can have this new feat select a Pick and apply the tag "Psionics.PsiFocus".  Thinking further maybe we should create a procedure instead that passes back if we are Psionic Focused or not. This way future things that affect Psi-Focus can be changed in just the one procedure.

So something like:
var v_focus as number
call PUPsiFocus
~ If not psionically focused get out now!
doneif (v_focus = 0) 

Procedure:
[code]
var v_focus as number
v_focus = hero.tagis[Hero.PsiFocus] + tagis[Psionics.PsiFocus]
[/code]

**Comment by shadowchemosh** [comment]
So I added the feat, the new tag, the new procedure.  But going through all the abilities and feats to use the new procedure to check for Psi Focus will need to wait.

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 20:14_

Maybe post by email will work for me...I had added feat and committed it but all of a sudden all connection to GitHub was failing so I couldn't push AND I couldn't post any messages past that last reply...so I guess we have both added the feat now. Lol

**Comment by shadowchemosh** [comment]
LOL Github was having technical issues for a bit.  So I ran into the same issue.  I did just get my stuff uploaded with https://github.com/shadowchemosh/HL-Pack-3PP/commit/e6973c931dec48ca38d4bbaeb92a48cf863a252a

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 20:21_

so now that we have both entered in the same feat...what do i do to not screw things up (and after i just got the last fiasco all sorted finally too!!)

**Comment by shadowchemosh** [comment]
I did modify the feat "Intuitive Fighting" as a Proof of Concept to make sure my idea works.  So we can use that feat as an example to enhance the other abilities.  But I am planning to release really soon so we can do that for next release next month.

Just do all the pushes up to GitHub.  It will do a merge.  So I will pull down the merge and fix and "duplicate" issues we get.  Then just pull down my changes afterwards.  not a big deal.  Just don't do that whole "un-commit" this time. =)

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 20:24_

lol but it sounded so simple!

i get this error when i try to push:
git -c diff.mnemonicprefix=false -c core.quotepath=false push -v --tags origin master:master
Pushing to https://psych777@github.com/shadowchemosh/HL-Pack-3PP.git
To https://psych777@github.com/shadowchemosh/HL-Pack-3PP.git
 = [up to date]      v1.0 -> v1.0
 = [up to date]      v1.1 -> v1.1
 = [up to date]      v1.2 -> v1.2
 = [up to date]      v1.4 -> v1.4
 = [up to date]      v1.5 -> v1.5
 = [up to date]      v3.00 -> v3.00
 = [up to date]      v3.1 -> v3.1
 = [up to date]      v3.2 -> v3.2
 = [up to date]      v3.3 -> v3.3
 = [up to date]      v3.4 -> v3.4
 = [up to date]      v3.5 -> v3.5
 = [up to date]      v3.6 -> v3.6
 ! [rejected]        master -> master (fetch first)
error: failed to push some refs to 'https://psych777@github.com/shadowchemosh/HL-Pack-3PP.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**Comment by shadowchemosh** [comment]
So do a "pull" with a merge...  Then do the push again.

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 20:29_

i do pull with that merge option checked and it results in this: (the uncommitted changes line did not exist before the pull)
![image](https://cloud.githubusercontent.com/assets/10572815/6887542/aa429c82-d65a-11e4-9b68-05146f26e799.png)
so still showing i need 3 pulls even though i did a pull...

**Comment by shadowchemosh** [comment]
I would try a pull one more time.  Its showing uncommited because it has created a new version of the Feat file.  Then try the push.

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 20:39_

ive pulled 3 more times and it doesnt change :o(

**Comment by shadowchemosh** [comment]
And what happens when you push?

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 20:56_

git -c diff.mnemonicprefix=false -c core.quotepath=false push -v --tags origin master:master
Pushing to https://psych777@github.com/shadowchemosh/HL-Pack-3PP.git
To https://psych777@github.com/shadowchemosh/HL-Pack-3PP.git
 = [up to date]      v1.0 -> v1.0
 = [up to date]      v1.1 -> v1.1

 = [up to date]      v1.2 -> v1.2
 = [up to date]      v1.4 -> v1.4
 = [up to date]      v1.5 -> v1.5
 = [up to date]      v3.00 -> v3.00
 = [up to date]      v3.1 -> v3.1
 = [up to date]      v3.2 -> v3.2
 = [up to date]      v3.3 -> v3.3
 = [up to date]      v3.4 -> v3.4
 = [up to date]      v3.5 -> v3.5
 = [up to date]      v3.6 -> v3.6
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'https://psych777@github.com/shadowchemosh/HL-Pack-3PP.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

**Comment by shadowchemosh** [comment]
Commit the changed/merged files.  Then see what it will let you do.

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 21:3_

that appears to have let me get the tree up to date...but i have a feeling it messed up the feats file...
![image](https://cloud.githubusercontent.com/assets/10572815/6887697/648b432e-d65f-11e4-9399-b63448128232.png)

**Comment by shadowchemosh** [comment]
I can fix that.

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 21:5_

we probably gave Ualaa a lot more conversation than he expected for a feat request. lol

**Comment by shadowchemosh** [comment]
LOL that is very true. =)

So I merged the two feats together.  I had forgotten about the Pre-Req so thanks for doing those.  And I pushed back to git.

**Comment by shadowchemosh** [comment]
_From @psych777 on March 29, 2015 21:9_

sounds good

**Comment by shadowchemosh** [comment]
_From @IronGolem on March 29, 2015 23:29_

I'm obviously very late to this party, but...

I think it would be helpful to add a "Maintain psionic focus" tag in abAction, just like we already have for abilities that require expending psionic focus.  Not only would that be helpful for editing in general, but it would also provide for an easy filter for this feat's menu selection; just look for any abilities that have abAction.PUManFocus (or whatever we call it).

**Comment by shadowchemosh** [comment]
How is that really different than the tag I made?

**Comment by shadowchemosh** [comment]
_From @IronGolem on March 30, 2015 3:1_

Different function, at least if I'm reading your comment and the code for Persistent Focus correctly.  If I'm not mistaken, this feat adds Psionics.PsiFocus to its chosen pick, to denote it as always being treated as if the hero was psi-focused.  Correct?

The tag I propose would simply denote an ability as requiring psi focus to be kept up in order to function - so it would be added to feats like Psionic Dodge and class specials like Eternal Warrior.  As for Persistent Focus in particular, it would restrict its selection to abilities that have that tag (so that it can't be applied to, say, Evasion or Lay on Hands).

Admittedly, this does require hunting down abilities that require psi-focus and adding the tag to them.  (Edit:  Then again, we already need to do this to add the new procedure, so this doesn't add much extra work after all).

**Comment by shadowchemosh** [comment]
<strong>The tag I propose would simply denote an ability as requiring psi focus to be kept up in order to function - so it would be added to feats like Psionic Dodge and class specials like Eternal Warrior. As for Persistent Focus in particular, it would restrict its selection to abilities that have that tag (so that it can't be applied to, say, Evasion or Lay on Hands).</strong>
Ahh yes I see now.  You are correct in that I would like to tag just those "Things" that this feat can be used with.  And I agree which is why I added the tag Psionics.FocusPick also.
https://github.com/shadowchemosh/HL-Pack-3PP/blob/master/COM_3PPPack_UltimatePsionics%20-%20Tags.1st#L20

This tag was what I planned to use to mark those Psionic Focus Things.  Granted I was in a hurry and it could use a better name.  I worry about using something like abAction.PUManFocus just because its so close to a tag type that LW could use or add.  Then we conflict and the packages fall into safe mode and I get nasty emails and dogs and cats living together. =)

To mimic LW they like the 'Helper" tag for this.  I added to the Basic Pack the "PackHelper.?" tag group to allow me to mimic that logic. 
https://github.com/shadowchemosh/HL-Pack-Basic/blob/master/COM_BasicPack%20-%20Tags.1st#L13

I am not sure if it falls into a PackHelper.? tag?  Or a PsiHelper.? tag group should be made?  I admit naming things is always fun!  What about <strong>PsiHelper.fPUPerFocu</strong> or just <strong>Psionics.fPUPerFocu</strong> and have it name="Persistent Focus Feat Usable"?

**Comment by shadowchemosh** [comment]
_From @IronGolem on March 30, 2015 23:19_

Idunno, that seems pretty narrow, as opposed to something that would have more widespread utility.  I see your point about potentially stepping on LW's toes with an abAction tag, but then why do we have one for expending focus?  It just seems to me that a tag saying "this thing requires psionic focus to be maintained" has a lot of potential use even outside of interacting with this one feat.

How about I write up a proof-of-concept tonight so we can all see what this would look like, and if it's problematic we can get rid of it?

**Comment by shadowchemosh** [comment]
Sure go for it. =)
